File: graph/2606/jinho.js
const fs = require('fs');
const input = fs.readFileSync('a.txt').toString().trim().split('\n'); 

const N = parseInt(input[0]);
const M = parseInt(input[1]);

const result = [1];
const matrix = [];

for (let i=0; i<N+1; i++){
    const arr = [];
    for (let j=0; j<N+1; j++){
        arr.push(0);
    }
    matrix.push(arr);
}

for (let i=2; i<M+2; i++){
    const [x, y] = input[i].split(' ').map(e => parseInt(e));
    matrix[x][y] = 1;
    matrix[y][x] = 1;
}

const bfs = (_matrix, idx) => {
    
    let done = true;
    const needsBfs = [];
    for (let i=0; i<N+1; i++){
        if (_matrix[idx][i] === 1 && result.indexOf(i) === -1){
            needsBfs.push(i)
            result.push(i);
            done = false;
        }
    }
    
    if (done === true){
        return;
    }   
    for (let i=0; i<needsBfs.length; i++){
        bfs(_matrix, needsBfs[i]);
    }
}

bfs(matrix, 1);
console.log(result.length -1);

//success

File: graph2/2178/jinho.py
from collections import deque

N, M = map(int, input().split())
matrix = []
result = []

for _ in range(N):
    matrix.append(list(map(int, input())))
    result.append([0] * M)

result[0][0] = 1
visited = []
def bfs(x, y):
    queue = deque()
    queue.append([x,y])
    while queue:
        x, y = queue.popleft()
        if result[N-1][M-1] > 1:
            return result[N-1][M-1]
        if 101*x + y not in visited:
            visited.append(101*x + y)
            if x+1<N and matrix[x+1][y]==1:
                result[x+1][y] = result[x][y] + 1
                # bfs_visited.append([x+1, y])
                queue.append([x+1, y])

            if y+1<M and matrix[x][y+1]==1:
                result[x][y+1] = result[x][y] + 1
                # bfs_visited.append([x, y+1])
                queue.append([x, y+1])
            
            if x>0 and matrix[x-1][y]==1:
                result[x-1][y] = result[x][y] + 1
                # bfs_visited.append([x-1, y])
                queue.append([x-1, y])

            if y>0 and matrix[x][y-1]==1:
                result[x][y-1] = result[x][y] + 1
                # bfs_visited.append([x, y-1])
                queue.append([x, y-1])

    return result[N-1][M-1]

print(bfs(0, 0))
